flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2988,6 +2998,10 @@ class ScalarExprLowering {
   }
   template <typename A>
   ExtValue genref(const A &a) {
+    if (inInitializer) {
+      // Initialization expressions can never allocate memory.
+      return genval(a);
+    }
```

Comment:
The problem here is we don't really want a reference to the constant value. We want the constant value itself (which is what genval does), and we will reinterpret it as a memory address. It's pretty straightforward in fact.

Reply:
Is there a reason we are creating a new SEL context here via `genExtAddrInInitializer`? I'm not sure that makes much sense, but I left it rather than tumbling down a slippery slope here.